File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys 

num,count=list(map(int,sys.stdin.readline().rstrip().split()))
lengths=list(map(int,sys.stdin.readline().rstrip().split()))

left=0
right=max(lengths)

while True:
    
    if left>right:#탈출 조건 !! 
        print(right)
        break
 
    mid=(left+right)//2
    sum=0

    for i in lengths:
        if i>mid:
            sum+=i-mid
    #print("나무 길이: ",mid)
    #print("총 길이: ",sum)

    if sum<count: #더 작게 잘라야함 
        right=mid-1
    
    elif sum>=count: # 더 크게 잘라야할수도 있어  
        left=mid+1

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;


public class Main {


    public static void main(String[] args) throws IOException {

        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(br.readLine()," ");
        int N=Integer.parseInt(st.nextToken());
        int M=Integer.parseInt(st.nextToken());
        int min,max,mid;
        long sum=0;
        int[] arr=new int[N];
        st=new StringTokenizer(br.readLine()," ");
        for(int i=0;i<N;i++) arr[i]=Integer.parseInt(st.nextToken());

        Arrays.sort(arr); // 오름차순 정렬
        min=0;
        max=arr[N-1];

        while(true){

            if(min>max) break;

            sum=0;
            mid=(min+max)/2;

            for(int i=0;i<N;i++){
                if(arr[i]>mid) sum+=arr[i]-mid;
            }
            if(sum>=M){ // 더 자르기
                min=mid+1;
            }
            else{ // 덜 자르기
                max=mid-1;
            }

        }
        System.out.println(max);

    }

}